### PR TITLE
curvefs/metaserver: fix s3compact crash when partition have been deleted

### PR DESCRIPTION
**Pull request by h0hmj**
we need erase s3compact from vector when partition is deleted

### CODE REVIEW

File: curvefs/src/metaserver/s3compact.h [hunk 1]
```diff
@@ -37,20 +36,21 @@ using curvefs::common::PartitionInfo;
...
 
     PartitionInfo GetPartition() {
-        return *partitionInfo_;
+        return partitionInfo_;
     }
 
  private:
     std::shared_ptr<InodeStorage> inodeStorage_;
-    PartitionInfo* partitionInfo_;
+    // we only need data that will not be changed after being created.
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact.h:52]: if so, make it const

File: curvefs/src/metaserver/s3compact.h [hunk 2]
```diff
@@ -37,20 +36,21 @@ using curvefs::common::PartitionInfo;
 class S3Compact {
  public:
     explicit S3Compact(std::shared_ptr<InodeStorage> inodeStorage,
-                       PartitionInfo* partitionInfo)
+                       PartitionInfo partitionInfo)
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact.h:39]: const PartitionInfo& partitionInfo

File: curvefs/src/metaserver/s3compact.h (hunk 1 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact.h:52]: done

File: curvefs/src/metaserver/s3compact.h (hunk 2 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact.h]: done

**Review by wu-hanqing — APPROVED** [review]

**Comment by h0hmj** [comment]
recheck

File: curvefs/src/metaserver/s3compact_manager.cpp [hunk 3]
```diff
@@ -183,16 +183,26 @@ void S3CompactManager::Stop() {
...
     if (s3compacts.empty()) {
+        // fastpath
         sleeper_.wait_for(std::chrono::milliseconds(opts_.enqueueSleepMS));
+        return;
     }
-    for (const auto& s3compact : s3compacts) {
+    std::vector<int> toErase;
+    for (int i = 0; i < s3compacts.size(); i++) {
+        auto s3compact(s3compacts[i].lock());
+        if (!s3compact) {
```
Comment (cw123) [inline curvefs/src/metaserver/s3compact_manager.cpp:200]: how to judge partition is deleted

Comment (cw123) [reply curvefs/src/metaserver/s3compact_manager.cpp:200]: when partition is under deleting, the compact record should be deleted

Comment (cw123) [reply curvefs/src/metaserver/s3compact_manager.cpp:200]: when the partition is deleting, it will reject update inode request.

Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_manager.cpp:200]: only partition hold the shared_ptr of s3compact, if partition is deleted, s3compact will be deleted.
weak_ptr lock() is thread safe.
if s3compact in partition has been deconstructed. lock() will return nullptr.
there is no deleting state, just deleted or not deleted.

**Review by cw123 — APPROVED** [review]

**Comment by h0hmj** [comment]
recheck

**Comment by YunhuiChen** [comment]
recheck

**Comment by h0hmj** [comment]
recheck

**Comment by h0hmj** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Pull request merged**